packages/autocomplete-js/src/render.ts

```diff
@@ -88,75 +90,68 @@ export function renderPanel<TItem extends BaseItem>(
   dom.panel.classList.toggle('aa-Panel--touch', isTouch);
   dom.panel.classList.toggle('aa-Panel--stalled', state.status === 'stalled');
 
-  const sourceElements = state.collections.map(({ source, items }) => {
-    const sourceElement = Element('section', { class: classNames.source });
-
-    if (source.templates.header) {
-      const headerElement = Element('div', { class: classNames.sourceHeader });
-
-      renderTemplate({
-        template: source.templates.header({
-          root: headerElement,
-          state,
-          source,
-          items,
-        }),
-        parent: sourceElement,
-        element: headerElement,
-      });
-    }
-
-    if (items.length > 0) {
-      const listElement = Element('ul', {
-        class: classNames.list,
-        ...propGetters.getListProps({
-          state,
-          props: autocomplete.getListProps({}),
-          ...autocompleteScopeApi,
-        }),
-      });
-      const listFragment = document.createDocumentFragment();
-
-      items.forEach((item) => {
-        const itemElement = Element('li', {
-          class: classNames.item,
-          ...propGetters.getItemProps({
+  const children = pragma('div', {
+    className: 'aa-PanelLayout',
+    children: state.collections.map(({ source, items }) => {
+      return pragma('section', {
+        className: classNames.source,
+        children: pragma('ul', {
+          className: classNames.list,
+          ...propGetters.getListProps({
             state,
-            props: autocomplete.getItemProps({ item, source }),
+            props: autocomplete.getListProps({}),
             ...autocompleteScopeApi,
           }),
-        });
-
-        renderTemplate({
-          template: source.templates.item({ root: itemElement, item, state }),
-          parent: listFragment,
-          element: itemElement,
-        });
-      });
-
-      listElement.appendChild(listFragment);
-      sourceElement.appendChild(listElement);
-    }
-
-    if (source.templates.footer) {
-      const footerElement = Element('div', { class: classNames.sourceFooter });
-      renderTemplate({
-        template: source.templates.footer({
-          root: footerElement,
-          state,
-          source,
-          items,
+          children: [
+            source.templates.header &&
+              pragma('div', {
+                className: classNames.sourceHeader,
+                children: [
+                  source.templates.header({
+                    pragma,
+                    pragmaFrag,
+                    items,
+                    source,
+                    state,
+                  }),
+                ],
+              }),
+            ...items.map((item) => {
+              return pragma('li', {
+                className: classNames.item,
```

Comment:
Do we need `key` prop here?

Reply:
Added in 120532c3d5c9eb6e5155610f1d2d1cc54bd33cd3 and 9345323653123d2f50269ccdf9e7ab7951cbaafc.